## Import libraries and dataset

In [335]:
import pandas as pd
pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 500)

In [336]:
path = './output.csv'

In [337]:
data = pd.read_csv(path)

###  Shape of the data

In [338]:
data.shape

(8184, 37)

### Features

In [339]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8184 entries, 0 to 8183
Data columns (total 37 columns):
LeasedFrom                 7901 non-null object
Engines                    8183 non-null object
CrashSiteElevation         8000 non-null object
Duration                   7930 non-null object
Narrative                  8130 non-null object
FlightNumber               8172 non-null object
Crew                       8184 non-null object
Location                   8184 non-null object
InvestigatingAgency        7930 non-null object
Type                       8182 non-null object
GroundCasualities          8097 non-null object
Status                     2590 non-null object
OnBehalfOf                 7950 non-null object
TotalFatalities            8184 non-null object
TotalAirFrameHrs           8165 non-null object
Nature                     8182 non-null object
AirplaneDamage             8184 non-null object
Released                   7930 non-null object
CarrierNumber              81

In [340]:
print(len(data.columns))
data.columns

37


Index(['LeasedFrom', 'Engines', 'CrashSiteElevation', 'Duration', 'Narrative',
       'FlightNumber', 'Crew', 'Location', 'InvestigatingAgency', 'Type',
       'GroundCasualities', 'Status', 'OnBehalfOf', 'TotalFatalities',
       'TotalAirFrameHrs', 'Nature', 'AirplaneDamage', 'Released',
       'CarrierNumber', 'Phase', 'Passengers', 'OperatedBy',
       'DestinationAirport', 'OperatingFor', 'Date', 'Time', 'Cycles',
       'DownloadReport', 'DepartureAirport', 'Issued', 'Number',
       'Registration', 'Operator', 'DurationOfInvestigation', 'AirplaneFate',
       'CollisionCasualties', 'FirstFlight'],
      dtype='object')

### Examine Nulls and Data types

In [341]:
# Check the null values in the dataset
# Function to check the null counts in a dataframe
def checknullcount(df):
    info = []
    columns = ('Feature', 'Nullcount', 'Dtypes')
    for col in df.columns:
        nullcount = df[col].isnull().sum()
        datatype = df[col].dtypes
        info.append([col, nullcount, datatype])
    return (pd.DataFrame(columns=columns, data=info).sort_values('Nullcount', ascending = False))

(checknullcount(data))

Feature  Nullcount   Dtypes
30                   Number       8184  float64
11                   Status       5594   object
33  DurationOfInvestigation        752   object
0                LeasedFrom        283   object
17                 Released        254   object
29                   Issued        254   object
8       InvestigatingAgency        254   object
3                  Duration        254   object
27           DownloadReport        254   object
21               OperatedBy        234   object
12               OnBehalfOf        234   object
2        CrashSiteElevation        184   object
35      CollisionCasualties        172   object
26                   Cycles        102   object
10        GroundCasualities         87   object
4                 Narrative         54   object
23             OperatingFor         40   object
14         TotalAirFrameHrs         19   object
5              FlightNumber         12   object
34             AirplaneFate         12   object
25                     Time         12   object
32                 Operator          3   object
9                      Type          2   object
15                   Nature          2   object
1                   Engines          1   object
31             Registration          0   object
28         DepartureAirport          0   object
18            CarrierNumber          0   object
24                     Date          0   object
22       DestinationAirport          0   object
20               Passengers          0   object
19                    Phase          0   object
16           AirplaneDamage          0   object
13          TotalFatalities          0   object
7                  Location          0   object
6                      Crew          0   object
36              FirstFlight          0   object

## Check Feature Category - Aircraft and Airline Details

### LeasedFrom
- The company, organisation or individual that owns the aircraft, but did not operate it at the time of the accident. 
    - Null count - 283, datatype = object
        - Here I can assume that when the 'Leased From'value is null, the operator is using their own aircraft.
    - What do I i do?
        - Drop the column LeasedFrom

In [342]:
checkcols = ['OnBehalfOf', 'OperatedBy', 'OperatingFor', 'Operator', 'LeasedFrom']
cnt = []
null = []

for col in checkcols:
    row = []
    row.append(data[col].nunique())
    row.append(data[col].isnull().sum())
    row.append('a')
    data[col].fillna('-', inplace = True)
    null.append(row)
cols = ['Nunique','NullCount','x']
pd.DataFrame(columns=cols, data=null, index = checkcols)

Nunique  NullCount  x
OnBehalfOf         10        234  a
OperatedBy          8        234  a
OperatingFor      183         40  a
Operator         2745          3  a
LeasedFrom         12        283  a

### OperatingFor
- The company, organisation or individual operating the aircraft at the time of the accident. This does not necessarily have to be the owner. 

In [343]:
data['OperatingFor'].value_counts()

-                                                      7908
United Express                                           11
United Nations - UN                                      10
Saudi Arabian Airlines                                    8
FedEx Express                                             8
Delta Connection                                          7
American Eagle                                            7
Royal Australian Air Force - RAAF                         6
USAir Express                                             4
US Airways Express                                        4
Qantas Empire Airways                                     3
Sri Lanka Air Force                                       3
Manx2                                                     3
Air Force of the Democratic Republic of the Congo         3
Air Méditerranée                                          3
Baron Aviation                                            3
Platinum Jet Management                 

### OnBehalfOf

In [344]:
data['OnBehalfOf'].value_counts()

-                                8173
U.S. Forest Service                 3
Cargo North                         1
Swiss International Air Lines       1
Continental Express                 1
US Department of State              1
Air France                          1
US Airways Express                  1
Northwest Airlink                   1
HOP!                                1
Name: OnBehalfOf, dtype: int64

### OperatedBy

In [345]:
data['OnBehalfOf'].value_counts()

-                                8173
U.S. Forest Service                 3
Cargo North                         1
Swiss International Air Lines       1
Continental Express                 1
US Department of State              1
Air France                          1
US Airways Express                  1
Northwest Airlink                   1
HOP!                                1
Name: OnBehalfOf, dtype: int64

### Operator

In [346]:
data['Operator'].value_counts()

United States Air Force - USAF                              462
-                                                           288
Royal Air Force - RAF                                       253
United States Navy                                          247
United States Army Air Force - USAAF                        221
Luftwaffe                                                    87
Air France                                                   62
Unknown                                                      49
Indian Air Force - IAF                                       47
Indian Airlines                                              46
Ejército del Aire                                            45
L'Armée de L'Air                                             42
Deutsche Lufthansa                                           40
Soviet Air Force                                             39
Eastern Air Lines                                            39
American Airlines                       

### These 4 seem the same. Let me look at them to know the difference
    - LeasedFrom has 7888 empty values, 283 null values, [10 unique values]
        - There are 13 rows with valid Leased from values. In these 3 cases Operator is not provided.
    - OnBehalfOf has 7939 empty values, 234 null values [10 unique values]
        - There are only 11 valid values. I think I am going to drop this column
        - Mostly 
    - OperatedBy has 7943 empty values, 234 null values [8 unique values]
        - When operated By is filled, Operator is not provided. I will merge Operated By and Operator
    - OperatingFor has 7868 empty values, 40 null values [183 unique values]
        - When operating for is filled, Operator is not provided. I will merge Operated For and Operator
    - Operator has only 3 null values [ 2745 unique values]
        - Most are null here. I will drop all 3 null values

In [347]:
test = data[['OnBehalfOf', 'OperatedBy', 'OperatingFor', 'Operator', 'LeasedFrom','Date']]
test['OnBehalfOf'].fillna('-',inplace = True)
                          
test[(test['OnBehalfOf'] != '-') & (test['Operator'] != '-')]
# test['OnBehalfOf'].value_counts()

/Users/roshanlulu/anaconda3/lib/python3.6/site-packages/pandas/core/generic.py:3295: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._update_inplace(new_data)


OnBehalfOf OperatedBy OperatingFor                   Operator  \
1527  U.S. Forest Service          -            -  Neptune Aviation Services   
1891  U.S. Forest Service          -            -  Neptune Aviation Services   
2238  U.S. Forest Service          -            -     Minden Air Corporation   
4134           Air France          -            -               TAME Ecuador   

     LeasedFrom                   Date  
1527          -  Saturday 26 June 2010  
1891          -     Sunday 3 June 2012  
2238          -    Sunday 15 June 2014  
4134          -   Monday 20 April 1998

#### Operating For and Operator are never empty together. I will merge these 2 values into Operator, assuming it is the same.

In [348]:
# Merge 
operator = []
for index, op in enumerate(data['Operator']):
    if op == '-':
        op = data.ix[index, 'OperatingFor']
    else:
        pass
    operator.append(op)

data['Operator'] = operator
# After merging there are still 12 columns which dont have operator information
# 12 is a small number so I choose to remvoe these rows!
# data['Operator'] = [col.replace(' ', '_') for col in ufo.columns]
print(len(data))
data = data[data['Operator'] != '-']
len(data)


8184


8172

#### Drop the other columns - LeasedFrom,OnBehalfOf, OperatedBy

In [349]:
# Drop the columns
drop_col = ['LeasedFrom', 'OperatedBy', 'OperatingFor', 'OnBehalfOf']
data.drop(drop_col, axis = 1, inplace = True)

In [350]:
data.shape

(8172, 33)

In [351]:
# No of Unique Operators - 2857

data['Operator'].nunique()
# data['Operator'].value_counts()
# I amnot changing the operator names now. Once I get the country cleaned up, Then I can group the operators

2857

In [352]:
# There are a lot of names here. I want to categorize the airforce ones together for now
# Aim is to get something like Airforce
# Im nto grouping the other lines, since they can be used to subclassify within a coutnry classification
# Remove any text with numbers in it

data['Operator'] = ['Aeroflot' if 'Aeroflot' in t 
                    else 'Air Force' if 'Air Force' in t
#                     else 'US Air Force' if 'USAF' in t or 'USAAF' in t or 'United States Navy' in t
                else t for t in data['Operator']]
print(data['Operator'].nunique())
data['Operator'].value_counts()
 # Use regular expression to remove any text with numbers


2748


Air Force                                                   1359
Aeroflot                                                     360
United States Navy                                           247
Luftwaffe                                                     87
Air France                                                    62
Unknown                                                       49
Indian Airlines                                               48
Ejército del Aire                                             45
L'Armée de L'Air                                              42
Deutsche Lufthansa                                            40
Eastern Air Lines                                             40
American Airlines                                             38
Avianca                                                       37
Imperial Airways                                              36
Delta Air Lines                                               34
KLM Royal Dutch Airlines 

### Nature
- Nature of the flight. For passenger flights ‘domestic’  or ‘international’ is added where known. 
For a full list of flight natures used, see Appendix 3

In [365]:
data.reset_index(inplace = True)
### Now might be a good time to check the nature of the flight. If that is categorized well
data['Nature'].nunique()
## There are 25 unique nature of flights

## Can this be cleaned further?
data['Nature'].value_counts()

Military                             2181
Domestic Scheduled Passenger         1514
Unknown                               958
Cargo                                 825
International Scheduled Passenger     548
-                                     310
Domestic Non Scheduled Passenger      308
Ferry/positioning                     283
Training                              213
Executive                             212
Int'l Non Scheduled Passenger         147
Test                                  111
Passenger                              97
Agricultural                           71
Private                                68
Scheduled Passenger                    67
Illegal Flight                         59
Ambulance                              38
Survey/research                        37
Parachuting                            34
Fire fighting                          24
Non Scheduled Passenger                20
Official state flight                  19
Demonstration                     

In [354]:
# Id like to categorize them as follows into 3 categories.
# 0 - Others/Misc
# 1 - Passenger Flights
# 2 - Military
# Adding a new column for Flight_Nature_Code
code_list = []
for row in range(0, len(data)):
    code = 0
    nature = data.ix[row, 'Nature']
    if 'Military' in nature:
        code = 2
    elif 'Passenger' in nature:
        code = 1
    code_list.append(code)
data['Nature_Code'] = code_list

### Type
- Manufacturer and exact model involved in the accident. For Boeing models for instance a (first) customer code is being used. The second and third digits indicate this code. Customer code for Boeing 747-206 is '06', which is KLM. 
Where possible the name of the manufacturer is used at the time the airplane in question was built. For example, the MD-11 remains “McDonnell Douglas MD-11” and does not become Boeing MD-11 or Boeing (McDD) MD-11.



    - Ive reduced it into 27 categories of Type.
    - The ones with lesser no of incidents have been grouped together to form a category called Others

In [355]:
# import re
# # remove string with digits in it
# def cleanstring(text):
#     # split by word
#     # check if the word has a number
#     # delete word with number
#     text = re.sub("\S*\d\S*", "", text).strip()
#     t = text
#     if len(text) > 0:
#         t = text.split()
#         t = t[0]
#     return t

# data['Type'] = [cleanstring(t) for t in data['Type']]

In [356]:
data['Type'].value_counts()
# data['Type'].nunique()

# There are a lot of names here. I want to categorize it by removing the numbers first.
# Aim si to get something like Airbus, Douglas, Antonov, Boeing
# Remove any text with numbers in it

data['Type'] = ['Douglas' if 'Douglas' in t 
                else 'Antonov' if 'Antonov' in t
                else 'Cessna' if 'Cessna' in t
                else 'Airbus' if 'Airbus' in t
                else 'Boeing' if 'Boeing' in t
                else 'Bombardier' if 'Bombardier' in t
                else 'Embraer' if 'Embraer' in t
                else 'Curtiss' if 'Curtiss' in t
                else 'Catalina' if 'Catalina' in t
                else 'Learjet' if 'Learjet' in t
                else 'Fairchild' if 'Fairchild' in t
                else 'Lockheed' if 'Lockheed' in t
                else 'Beechcraft' if 'Beechcraft' in t
                else 'Britten-Norman' if 'Britten-Norman' in t
                else 'Jetstream' if 'Jetstream' in t
                else 'Ford' if 'Ford' in t
                else 'Yakovlev' if 'Yakovlev' in t
                else 'de Havilland' if 'de Havilland' in t
                else 'Sabreliner' if 'Sabreliner' in t
                else 'Lisunov' if 'Lisunov' in t
                else 'Tupolev' if 'Tupolev' in t
                else 'Junkers' if 'Junkers' in t
                else 'Vickers' if 'Vickers' in t
                else 'Fokker' if 'Fokker' in t
                else 'Swearingen' if 'Swearingen' in t
                else 'Ilyushin' if 'Ilyushin' in t
                else 'Others' for t in data['Type']]
print(data['Type'].nunique())
data['Type'].value_counts()
 # Use regular expression to remove any text with numbers


27


Douglas           1859
Others            1441
Antonov            635
Boeing             613
Lockheed           502
de Havilland       356
Curtiss            268
Cessna             268
Beechcraft         232
Junkers            228
Britten-Norman     183
Vickers            171
Fokker             152
Catalina           151
Ilyushin           145
Fairchild          139
Airbus             135
Learjet            127
Tupolev             99
Ford                98
Swearingen          81
Embraer             78
Lisunov             72
Yakovlev            57
Sabreliner          41
Jetstream           26
Bombardier          15
Name: Type, dtype: int64

In [357]:
## Id like to plot the Aircraft distributions wrt timelines
## See the dictibution of aircraft type to airline operator

In [358]:
# import re


# s = "ABCD abcd AB55 55CD A55D 5555"
# re.sub("\S*\d\S*", "", s).strip()

### FlightNumber - Verdict: Make it my unique ID.
- Flightnumber as assigned to the flight by the operator. They ar esupposed ot be unique values
- I am going to randomly assign unique values to it. I wont be retaining the original numbers. It will not make a difference to my data

In [364]:
# data[['CarrierNumber', 'FlightNumber']]
unique_id = [t for t in range(0, len(data))]
data['FlightNumber'] = unique_id

###  CarrierNumber  - Verdict: Drop it!

In [360]:
# I dont think carrier number is a useful feature.
# Im going to drop the column
data.drop('CarrierNumber', axis = 1, inplace = True)

### 37. FirstFlight - Is this the Year Built?

In [370]:
data['FirstFlight'].isnull().sum()
# No null values
data['FirstFlight'].value_counts()
# There are 2011 rows that have empty data
# Cleaning required
# sme have only year
# Some have the date of first flight and the age of the flight at that time
# Check the date of incident and check what we need for this field. 
# - Age mightbe a good field to get from here and the dat field

                                       2011
 1944                                   662
 1943                                   339
 1945                                   331
 1942                                   155
 1968                                   122
 1966                                   103
 1967                                    99
 1981                                    96
 1980                                    93
 1969                                    92
 1977                                    87
 1976                                    85
 1975                                    81
 1958                                    80
 1957                                    79
 1978                                    78
 1959                                    76
 1973                                    75
 1979                                    73
 1972                                    71
 1956                                    70
 1970                           

### 32. Registration
- The registrationmark applied on the  the aircaft at the time of the accident. Airplanes involved in some occurrences carried illegal/unofficial registration marks. In these cases the unofficial registration is given.

### 31. Number
- Manufacturer serial number/construction number. Some manufacturers like Boeing and McDonnell Douglas include the line number, preceded by a /. Thus a Boeing 767, c/n 24542/28 means it has construction number 24542, and that it is the 28th B767 off the production line. Construction numbers for each manufacturer are usually unique.

### 2.Engines
- Number and type (model and mark) of engines.


## Feature Category Fatalities

### 7. Crew
- Exact number of flight- and cabincrew members aboard the aircraft at the time of departure, and number of passengers fatally injured as a direct result of the accident. An injury is classified as fatal if death results from the injury within 30 days from the date of the accident (ICAO) Deadheading crewmembers are judged 'passengers'.


### 14. TotalFatalities

### 21. Passengers
- Number of passengers aboard the aircraft at the time of departure, and number of passengers fatally injured as a direct result of the accident. 
An injury is classified as fatal if death results from the injury within 30 days from the date of the accident (ICAO)
Total:	Number of occupants (crew + passengers) aboard the aircraft at the time of the accident, and number of occupants fatally injured as a direct result of the accident. An injury is classified as fatal if death results from the injury within 30 days from the date of the accident (ICAO).


### 11. GroundCasualities
- Number of fatalities on the ground as a direct result of the occurrence. Ground casualties exclude fatalities as a result of an airplane collision on the ground.
See also: Collision casualties


### 36. CollisionCasualties
- Number of fatalities aboard other aircraft involved in collision accidents.

## Feature Category Crash Details

### 3. CrashSiteElevation

### 20. Phase
- Phase of flight. The flight phases are in accordance with the broad phases defined by the CAST/ICAO Common Taxonomy Team (CICTT), as laid down in the Phase of Flight Definitions document (October, 2002)

### 4. Duration - Same as Total airframe hours?

### 5. Narrative
- Description of the occurrence.


### 25. Date
- Date of occurrence (local time), format DD MMM YYYY. DD or MMM becomes XX or XXX if the exact day or month of an occurrence is unknown.

### 29. DepartureAirport
- Last airport of departure before the accident (between brackets if known the IATA airport code).

### 23. DestinationAirport
- Scheduled destination airport. In case of diversions, the intended destination airport is given (between brackets if known the IATA airport code).


### 17. AirplaneDamage
- Describes the amount of damage to the airplane as a result of the occurrence. Possible values are: Written off (See also: Definitions), Substantial, Minor, None, and Unknown.


### 26. Time
- Time of the accident – local time.

### 35. AirplaneFate
- Describes the amount of damage to the airplane as a result of the occurrence. Possible values are: Written off (See also: Definitions), Substantial, Minor, None, and Unknown.


### 15. TotalAirFrameHrs - What is this?
- Total airframe hours at the time of the accident.

### 27. Cycles - What category to put this ?
- Total number of take-offs and landings at the time of the accident.


### 8. Location
- Location of the incident

## Feature Category - Investigation results

### 28. DownloadReport - What is this?

### 30. Issued -When was report issued?

### 18. Released

### 12. Status
- Status of the information. Final means that the official final investigation report was used to compile the accident description. Preliminary- Official means that the official preliminary investigation report was used and Preliminary means press information and other unofficial information was used.


### 34. DurationOfInvestigation

### 9. InvestigatingAgency

In [ ]:
data['Date'] = [d.split()[-1] for d in data['Date']]

In [ ]:
data = data[data['Date'] != 'unk']

In [ ]:
data = data[data['Date'] != 'unk.']

In [ ]:
data = data[data['Date'] != '195.']

In [ ]:
data = data[data['Date'] != '197x']

In [ ]:
import re

In [ ]:
test  = 'ABs9'
print(re.match("[0-9]", test))

In [ ]:
import re

data['Date'] = [0 if (re.match("[0-9]", d)) == None else int(d) for d in data['Date']]

In [ ]:
# data = data[data['Date'] > 1980]

In [ ]:
data.shape

In [ ]:
checknullcount(data)

In [ ]:
data['Nature'].value_counts()